prometheus_hardware_exporter/collector.py

```diff
@@ -874,22 +874,127 @@ def specifications(self) -> List[Specification]:
         ]
 
     def fetch(self) -> List[Payload]:
         """Load redfish data."""
-        payloads = []
-        service_status = self.redfish_helper.discover()
-        payloads.append(Payload(name="redfish_service_available", value=float(service_status)))
-
-        sensor_data = self.redfish_helper.get_sensor_data()
-        if not sensor_data:
-            logger.error("Failed to get sensor data via redfish.")
-            payloads.append(Payload(name="redfish_call_success", value=0.0))
-            return payloads
+        payloads: List[Payload] = []
+        with RedfishHelper(self.config) as redfish_helper:
+            service_status = redfish_helper.discover()
+            payloads.append(Payload(name="redfish_service_available", value=float(service_status)))
+
+            processor_count: Dict[str, int]
+            processor_data: Dict[str, List]
+            storage_controller_count: Dict[str, int]
+            storage_controller_data: Dict[str, List]
+            storage_drive_count: Dict[str, int]
+            storage_drive_data: Dict[str, List]
+            memory_dimm_count: Dict[str, int]
+            memory_dimm_data: Dict[str, List]
+
+            sensor_data: Dict[str, List] = redfish_helper.get_sensor_data()
+            processor_count, processor_data = redfish_helper.get_processor_data()
+            (
+                storage_controller_count,
+                storage_controller_data,
+            ) = redfish_helper.get_storage_controller_data()
+            network_adapter_count: Dict[str, Any] = redfish_helper.get_network_adapter_data()
+            chassis_data: Dict[str, Dict] = redfish_helper.get_chassis_data()
+            storage_drive_count, storage_drive_data = redfish_helper.get_storage_drive_data()
+            memory_dimm_count, memory_dimm_data = redfish_helper.get_memory_dimm_data()
+            smart_storage_health_data: Dict[
+                str, Any
+            ] = redfish_helper.get_smart_storage_health_data()
+
+        metrics = {
+            "sensor_data": sensor_data,
+            "processor_data": processor_data,
+            "processor_count": processor_count,
+            "storage_controller_data": storage_controller_data,
+            "storage_controller_count": storage_controller_count,
+            "network_adapter_count": network_adapter_count,
+            "chassis_data": chassis_data,
+            "storage_drive_data": storage_drive_data,
+            "storage_drive_count": storage_drive_count,
+            "memory_dimm_data": memory_dimm_data,
+            "memory_dimm_count": memory_dimm_count,
+            "smart_storage_health_data": smart_storage_health_data,
+        }
+
+        for metric_name, metric_data in metrics.items():
+            if not metric_data:
+                logger.error("Failed to get %s via redfish", metric_name)
 
         payloads.append(Payload(name="redfish_call_success", value=1.0))
-        for chassis_name, curr_sensor_data in sensor_data.items():
+
```

Comment:
When will we set redfish_call_success as 0?

Reply:
That's a good question. We're setting `redfish_service_available` to 0 in case redfish discover method doesn't find any redfish services. But `redfish_call_success` is very vague. Which call does it refer to since there's a lot of API calls that we're making to redfish? Maybe we can check a few important metrics and if any of them doesn't return anything, we can set it to 0. But I'm also not 100% sure about this.

Reply:
Please provide some proposals.

Reply:
After putting some I think I got the right proposal for this.

`redfish_call_success` should be mapped to the result for the initial redfish session login. In case the login fails, there's no point calling any of the metrics. 

So I've added the code below so that in case the `_get_redfish_obj` method returns `None` (which happens if the login fails due to an exception), then `redfish_call_success` is 0.

```python
        with RedfishHelper(self.config) as redfish_helper:
            # redfish_obj returns None if login was not successful
            if redfish_helper.redfish_obj is None:
                payloads.append(Payload(name="redfish_call_success", value=0.0))
                return payloads
            
           (...)
```